_posts/2021-09-15-drain-cleaner.md

```diff
@@ -0,0 +1,190 @@
+
+The Drain Cleaner must be deployed at the Kubernetes cluster level using a cluster-admin user, as we need to create a 
+dedicated service account, cluster role and binding to allow reading pods resources. A single instance is able to handle 
+all Kafka instances hosted on that Kubernetes cluster.
+
+With the validating webhook configuration in place, the Drain Cleaner is notified as soon as something tries to evict 
+Kafka or ZooKeeper pods and it annotates them with `strimzi.io/manual-rolling-update`. This tells the Cluster Operator 
+that thy need to be restarted in the next reconciliation phase, making sure the Kafka cluster is always available. 
+Restart annotation is not created if already present (idempotency) or the eviction request is running in dry-run mode.
+Pod restart will be delayed if that would cause one or more partitions to be under-replicated. 
+
+Once you have a Kafka cluster and the Drain Cleaner up and running, you can test it by draining one node hosting Kafka 
+and/or Zookeeper pods with the following command:
+
+```sh
+kubectl drain $NODE_NAME --delete-emptydir-data --ignore-daemonsets --timeout=6000s --force
+```
+
+Looking at the Drain Cleaner's logs, you will see the eviction events being notified for all pods on that node. 
+Only Kafka and Zookeeper pods are actually annotated for restart, all the rest is filtered out.
+
+```
+2021-08-03 08:07:00,218 INFO  [io.str.ValidatingWebhook] (executor-thread-17) Received eviction webhook for Pod my-cluster-zookeeper-2 in namespace test
+2021-08-03 08:07:00,218 INFO  [io.str.ValidatingWebhook] (executor-thread-17) Pod my-cluster-zookeeper-2 in namespace test will be annotated for restart
+2021-08-03 08:07:00,301 INFO  [io.str.ValidatingWebhook] (executor-thread-17) Pod my-cluster-zookeeper-2 in namespace test found and annotated for restart
+
+2021-08-03 08:07:01,508 INFO  [io.str.ValidatingWebhook] (executor-thread-17) Received eviction webhook for Pod my-cluster-kafka-0 in namespace test
+2021-08-03 08:07:01,508 INFO  [io.str.ValidatingWebhook] (executor-thread-17) Pod my-cluster-kafka-0 in namespace test will be annotated for restart
+2021-08-03 08:07:01,540 INFO  [io.str.ValidatingWebhook] (executor-thread-17) Pod my-cluster-kafka-0 in namespace test found and annotated for restart
+```
+
+Given that the node being evacuated is cordoned (unschedulable), pods restart operated by the Cluster Operator in the 
+next reconciliation phase will actually mov them away from that node.
+
+```
+2021-08-03 08:07:13 INFO  PodOperator:68 - Reconciliation #13(timer) Kafka(test/my-cluster): Rolling pod my-cluster-zookeeper-2
+2021-08-03 08:08:06 INFO  PodOperator:68 - Reconciliation #13(timer) Kafka(test/my-cluster): Rolling pod my-cluster-kafka-0
+2021-08-03 08:08:53 INFO  AbstractOperator:500 - Reconciliation #13(timer) Kafka(test/my-cluster): reconciled
+```
+
+Once you are finished, don't forget to make that node schedulable again:
+
+```sh
+kubectl uncordon $NODE_NAME
+```
```

Comment:
Should we have this here? You do this with playing with the Drain Cleaner. But in real life you would not uncordon it but go either upgrade it or just delete it etc. So I think we should either remove it or add some note that _in case you were just trying it out, you should uncordon it_.